Revamped counter example

ipopescu opened the pull request:
I have updated the counter contract and tests so we can use them in training and the mini-video series we want to add to the docs.

The new structure follows the instructions in the documentation and uses the latest test framework capabilities.

Related ticket for creating the videos: https://github.com/CasperLabs/ProServices_General/issues/76

In future PRs, I plan on using this code base also to demonstrate:
- upgrades and tests for an upgrade
- simple session code that doesn't call a contract

Tests run successfully:

<img width="1172" alt="Screen Shot 2022-09-20 at 3 49 58 PM" src="https://user-images.githubusercontent.com/4185994/191277623-8d088bfa-b442-4631-abfe-5c2eb605dd90.png">

ramujin commented on contract/src/main.rs:
```diff
@@ -46,51 +46,54 @@ pub extern "C" fn counter_get() {
     let (stored_contract_hash, _) =
-        storage::new_locked_contract(counter_entry_points, Some(counter_named_keys), None, None);*/
-
-    // Hash of the installed contract will be reachable through named keys
-    runtime::put_key(COUNTER_KEY, stored_contract_hash.into());
+        storage::new_locked_contract(counter_entry_points, Some(counter_named_keys), None, None); */
```
Maybe move this up next to the `storage::new_contract(...)` call so that someone doesn't inadvertently swap them in trying to create a locked contract and then receive errors because the `put_key` command is preceding it.

ipopescu replied:
Done!

ramujin reviewed (commented):
Looks good! I just left one minor comment in the contract code, but I didn't review the testing code since we are waiting for Karan to go over it anyway.

darthsiroftardis commented on contract/Cargo.toml:
```diff
@@ -14,7 +13,3 @@ path = "src/main.rs"
 doctest = false
 test = false
-
-[profile.release]
```
Lets preserve this please as it reduced Wasm size

darthsiroftardis commented on contract/src/main.rs:
```diff
@@ -46,51 +46,54 @@ pub extern "C" fn counter_get() {
 
-    // Create initial named keys of the contract.
+    // In the named keys of the contract, add a key for the count
     let mut counter_named_keys: BTreeMap<String, Key> = BTreeMap::new();
```
```suggestion
    let mut counter_named_keys = NamedKeys::new();
```

darthsiroftardis commented on contract/src/main.rs:
```diff
@@ -46,51 +46,54 @@ pub extern "C" fn counter_get() {
-    // Hash of the installed contract will be reachable through named keys
-    runtime::put_key(COUNTER_KEY, stored_contract_hash.into());
+    // Create a named key for the contract hash
+    runtime::put_key("counter", stored_contract_hash.into());
```
If we can redefine the "counter" here as a const at the top of the file for the sake of consistency

darthsiroftardis reviewed (commented):
A few small nits but overall looks really really good for a revamp

ipopescu commented:
Thanks for the review, @darthsiroftardis, and for catching those details. I've addressed your feedback.
I've also added a general "authors" contact as a link to Discord, instead of using individual emails:
> authors = ["CasperLabs <https://discord.com/invite/Q38s3Vh>"]
